### ISSUE TITLE
Erroneous CA1508

### ISSUE THREAD

**Issue by tamlin-mike** [dotnet/roslyn-analyzers#3468]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8

It seems the DFA for 1508 is all over the place.

```C#
// Just a bunch of tests for Roslyn Analyzers
using System;
using System.IO;

namespace RoslynAnalyzersBugtest
{
  public static class AnalyzerTests
  {
    public static void ErroneousCA1508_1(bool b1)
    {
      string s = ""; // CA1508 only works as expected if initializing 's' to null.
      if (b1) { s = "something"; }
      if (!string.IsNullOrEmpty(s)) // incorrectly flags IsNullOrEmpty
      {
        Console.WriteLine(s);
      }
    }

    public static void ErroneousCA1508_2()
    {
      // Error: CA1508  'stream = ...' is never 'null'. Ehhh, and...?
      using (var stream = new FileStream("blah", FileMode.OpenOrCreate))
      {
        // same analyzer bug here
        using (var reader = new StreamReader(stream))
        {
          // probable cause: The analyzer looks at "text != null", but reports for something *completely* different.
          for (string text = reader.ReadLine(); text != null; text = reader.ReadLine())
          {
          }
        }
      }
    }

    // Probably same bug as in ErroneousCA1508_2, only here the loop condition is greatly simplified
    public static void ErroneousCA1508_3(int n)
    {
      { // intentional scoping
        using (var bw = new BinaryWriter(new MemoryStream())) { }
      }
      { // intentional scoping
        // the following loop generates erroneous CA1508, not only for
        // the code inside the loop, but also in the _previous scope_.
        for (int i = 0; i < n; i++)
        {
          using (var bw = new BinaryWriter(new MemoryStream())) { }
        }
      }
    }
  }
}
```

**Comment by mavasani** [comment]
Just checked on `3.0.0-beta3.final` and the only false violation/false positive is on `if (!string.IsNullOrEmpty(s)) // incorrectly flags IsNullOrEmpty`. Others seem to be fixed.

**Comment by tamlin-mike** [comment]
Good news, mavasani. Glad the repro still contained something useful.

**Comment by sungam3r** [comment]
My case for false positive:
```c#
public interface IAuditWriter
{
  void Write(string operationName, string message, params object?[] propertyValues);
}

public sealed class DebugAuditWriter : IAuditWriter
{
  public void Write(string operationName, string message, params object?[] propertyValues)
  {
    Debug.WriteLine($"[Audit] {operationName}: {message}");

    if (propertyValues != null) // Warning	CA1508	'propertyValues != null' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
    {
      foreach (object? value in propertyValues)
        Debug.WriteLine($"\t{value}");
    }
  }
}

public void Primitive_AuditWriter_Should_Work(string operationName, string message, object[] propertyValues)
{
  new DebugAuditWriter().Write("dfd", "dfdf", null); // false positive here
  new DebugAuditWriter().Write("dfd", "dfdf", null, null);
  new DebugAuditWriter().Write("dfd", "dfdf", 1);
}

**Comment by mavasani** [comment]
@tamlin-mike Only remaining issue from your original post is a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/3685, for which I have an open PR. I'll close this issue.

@sungam3r Your case seems to same as https://github.com/dotnet/roslyn-analyzers/issues/3845
